src/error.rs

```diff
@@ -28,51 +30,139 @@ macro_rules! write_err {
 
 /// Hex decoding error.
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub enum HexToBytesError {
-    /// Non-hexadecimal character.
-    InvalidChar(InvalidCharError),
+pub enum FromHexError<E> {
+    /// Failed to create object from bytes iterator.
+    Invalid(E),
     /// Purported hex string had odd length.
     OddLengthString(OddLengthStringError),
 }
 
-impl fmt::Display for HexToBytesError {
+impl<E: fmt::Display> fmt::Display for FromHexError<E> {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        use HexToBytesError::*;
+        use FromHexError::*;
 
         match *self {
-            InvalidChar(ref e) => write_err!(f, "invalid char, failed to create bytes from hex"; e),
+            Invalid(ref e) => write_err!(f, "invalid"; e),
             OddLengthString(ref e) =>
                 write_err!(f, "odd length, failed to create bytes from hex"; e),
         }
     }
 }
 
 #[cfg(feature = "std")]
-impl std::error::Error for HexToBytesError {
+impl<E: std::error::Error + 'static> std::error::Error for FromHexError<E> {
     fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
-        use HexToBytesError::*;
+        use FromHexError::*;
 
         match *self {
-            InvalidChar(ref e) => Some(e),
+            Invalid(ref e) => Some(e),
             OddLengthString(ref e) => Some(e),
         }
     }
 }
 
-impl From<InvalidCharError> for HexToBytesError {
+impl<E> From<OddLengthStringError> for FromHexError<E> {
     #[inline]
-    fn from(e: InvalidCharError) -> Self { Self::InvalidChar(e) }
+    fn from(e: OddLengthStringError) -> Self { Self::OddLengthString(e) }
+}
+
+/// Error decoding a hex string that explicitly excludes a prefix.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum FromNoPrefixHexError<E> {
+    /// Input string contains a prefix.
+    Prefix(ContainsPrefixError),
```

Comment:
In principle this shouldn't be needed. A `0x` prefix is just invalid digit `x`. It is true that handling the special case can be helpful in error messages but we have the information in the input string, no need to store it (and in case of no-alloc we can store a `bool`).

Reply:
Agreed but, at least in my experience, getting the invalid character error is a bit jarring, contains prefix error is more informative. Since this is a hopefully write-once-change-very-little crate I thought the thoroughness was worth it.

Reply:
I've pulled all the prefix logic out of this PR. Reducing the scope to just the error stuff.

Reply:
As I said, you don't need a separate variant for it.

Reply:
Also marking as resolved because of the relevance to prefixes.